gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/SnappyActivation.java

```diff
@@ -311,13 +311,17 @@ public void setResultDescription(GenericResultDescription resultDescription) {
 
   public static void executeOnLeadNode(SnappySelectResultSet rs, GfxdResultCollector<Object> rc,
       String sql, boolean enableStreaming, long connId, String schema, ParameterValueSet pvs,
-      boolean isPreparedStatement, boolean isUpdateOrDelete)
+      boolean isPreparedStatement, boolean isUpdateOrDelete, LanguageConnectionContext lcc)
       throws StandardException {
     // TODO: KN probably username, statement id and connId to be sent in
     // execution and of course tx id when transaction will be supported.
     LeadNodeExecutionContext ctx = new LeadNodeExecutionContext(connId);
     LeadNodeExecutorMsg msg = new LeadNodeExecutorMsg(sql, schema, ctx, rc, pvs,
         isPreparedStatement, false, isUpdateOrDelete);
+    // release all locks before sending the message else it can lead to deadlocks
+    if (lcc != null) {
+      lcc.getTransactionExecute().releaseAllLocks(true, true);
```

Comment:
If its just about routing of DDLs, you may also use returnRows flag (as false) to identify DDLs. This way release of locks would also not required for prepare phase (line no 341).

While reviewing for this saw some discrepancy in usage of cc.markAsDDLForSnappyUse and cc.isMarkedAsDDLForSnappyUse. I will work on its cleanup.